win32ss/gdi/eng/stretchblt.c

```diff
@@ -397,10 +534,169 @@ IntEngStretchBlt(SURFOBJ *psoDest,
+        /* if cySrc < 0 and cyDest <0 (both negative, we change both their signs  */
+        /* because the outcome is the same as if they were both positive and we   */
+        /* reverse both their top and bottom values                               */
+        if ((cySrc < 0) && (cyDest < 0))
+        {
+            lTmp = DestRect->top;
+            DestRect->top = DestRect->bottom;
+            DestRect->bottom = lTmp;
+            lTmp = SourceRect->top;
+            SourceRect->top = SourceRect->bottom;
+            SourceRect->bottom = lTmp;
+            cySrc = -cySrc;
+            cyDest = -cyDest;
+        }
+
+        /* if the cxDest >= 0 and cxSrc < 0 then we reverse  */
+        /* the signs of both cxDest and cxSrc                */
+        if ((cxDest >= 0) && (cxSrc < 0))
+        {
+            cxDest = -cxDest;
+            cxSrc = -cxSrc;
+        }
+
+        /* if the cxDest < 0 and cxSrc >= 0 then we reverse */
+        /* the left and right values                        */
+        else if ((cxDest < 0) && (cxSrc >= 0))
+        {
+            lTmp = SourceRect->left;
+            SourceRect->left = SourceRect->right;
+            SourceRect->right = lTmp;
+        }
+
+        /* if the cyDest >= 0 and cySrc < 0 then we reverse  */
+        /* the signs of both cxDest and cxSrc                */
+        if ((cyDest >= 0) && (cySrc < 0))
+        {
+            cyDest = -cyDest;
+            cySrc = -cySrc;
+        }
+
+        /* if the cyDest < 0 and cySrc >= 0 then we reverse */
+        /* the top and bottom values                        */
+        else if ((cyDest <= 0) && (cySrc > 0))
+        {
+            lTmp = SourceRect->top;
+            SourceRect->top = SourceRect->bottom;
+            SourceRect->bottom = lTmp;
+        }
+    }
+
+    /*****************************************************************************************************************************
+      We want to send the flip state to IntEngBitBlt without changing its parameter list.
+      So a way that we can do this is to use the DestRect to carry this information along with it.
+      Since there are four values, we can use their relative positions (coordinates) to indicate the four flip conditions.
+      if the cx = 0 then there can be no x-flip and if cy = 0 there can be no y-flip. We have no concerns about these conditions.
+      So we can set the four flip conditions as follows:
+
+      flip = 0 means no flips                             therefore left < right and top < bottom (normal well-formed rect)
+      flip = 1 means that we are flipped about the y-axis therefore left < right and top > bottom
+      flip = 2 means that we are flipped about the x-axis therefore left > right and top < bottom
+      flip = 3 means that we are flipped about both axes  therefore left > right and top > bottom
+    *****************************************************************************************************************************/
```

Comment:
@tkreuzer and @jimtabor Do either of you have any comments about my explanation above on forwarding this information? This was completely my idea and I am open to better suggestions. Thanks. Doug Lyons

Reply:
See my comment in `win32ss/gdi/dib/dib1bpp.c`

Reply:
Using Destination Rectangle where left > right and top > bottom to forward flip info in addition to BMF_TOPDOWN bit for SourceSurface.fjBimap flags for Top-to-Bottom flag from Lazarus and PeaZip. This should work well.